### DISCUSSION TITLE
DHCP service needs some love!  ;)

### DISCUSSION THREAD

**Discussion by t1937r0** [TechnitiumSoftware/DnsServer#348] [Ideas]
First, let me say thanks for the DnsServer project.  You've done a great job including features that make sense and are easy to set up using docker-compose.

One feature that I'd like to request is the ability to edit a lease reservation and change the assigned IP address.  Specifically to an address on the network defined in the scope, but outside of the scope IP address range.  I like to sort and group my network devices into smaller ranges by device type, but it is not easy to do so if the only option is to accept the address dynamically assigned prior to converting the lease to a reservation.  I typically use the scope range for dynamc IPs for mobile devices and other itinerant devices.

So, for example, I'd like to have the following in a scope defined for the 10.0.2.0/24 network:

The dynamic IPs (generally transient and/or non-family mobile devices) would be in the 10.0.2.200 - 10.0..2.250 range

The following would all be lease reservation ranges

  - servers and storage in the 10.0.2.10 - 10.0.2.19 range
  - printers and scanners in the 10.0.2.20 - 10.0.2.29 range
  - octoprint (3d printer) instances in the 10.0.2.30 - 10.0.2.49 range
  - family PCs in the 10.0.2.50 - 10.0.2.59 range
  - family laptops in the 10.0.2.60 - 10.0.2.69 range
  - family mobile devices in the 10.0.2.70 - 10.0.2.89 range
  - work laptops in the10.0.2.100 - 10.0.2.109 range

The "lease reservation" ranges will have unused IPs that I don't want to have dynamically assigned.

If I'm just being a control freak, go ahead and say so.  On the other hand, I think this feature could be very useful.

Thanks for listening!

Robin

**Comment by ShreyasZare** [comment]
> First, let me say thanks for the DnsServer project. You've done a great job including features that make sense and are easy to set up using docker-compose.

Thanks for the compliments. That is the one of the main aims of the project to make complex installation/configuration part easy to make it less intimidating to people who are not experts, and for experts to make it quick and easy saving time.
 
> One feature that I'd like to request is the ability to edit a lease reservation and change the assigned IP address. Specifically to an address on the network defined in the scope, but outside of the scope IP address range. I like to sort and group my network devices into smaller ranges by device type, but it is not easy to do so if the only option is to accept the address dynamically assigned prior to converting the lease to a reservation. I typically use the scope range for dynamc IPs for mobile devices and other itinerant devices.

Assigning reserved lease outside of scope is not feasible as it would break the Scope design. It would be better if you can create multiple scopes out of which one will use dynamic allocations and the rest will only allocate reserved leases that are configured.

You can change the reserved lease's IP address by editing the scope. The device will automatically get the new IP address allocated when it tries to renew the existing lease, reboots, or connects to the network again.

> So, for example, I'd like to have the following in a scope defined for the 10.0.2.0/24 network:
> 
> The dynamic IPs (generally transient and/or non-family mobile devices) would be in the 10.0.2.200 - 10.0..2.250 range
> 
> The following would all be lease reservation ranges
> 
>     * servers and storage in the 10.0.2.10 - 10.0.2.19 range
> 
>     * printers and scanners in the 10.0.2.20 - 10.0.2.29 range
> 
>     * octoprint (3d printer) instances in the 10.0.2.30 - 10.0.2.49 range
> 
>     * family PCs in the 10.0.2.50 - 10.0.2.59 range
> 
>     * family laptops in the 10.0.2.60 - 10.0.2.69 range
> 
>     * family mobile devices in the 10.0.2.70 - 10.0.2.89 range
> 
>     * work laptops in the10.0.2.100 - 10.0.2.109 range
> 
> 
> The "lease reservation" ranges will have unused IPs that I don't want to have dynamically assigned.

If you don't want to have a scope for each of the device groups, just use the Exclusion option in the Scope to exclude the IP address ranges that must not be dynamically allocated. So you keep the range free from dynamic allocation to be available for reserved leases.

> If I'm just being a control freak, go ahead and say so. On the other hand, I think this feature could be very useful.

This happens with many people. Even I keep a few devices on my network with specific IP address just in case there is issue with the DNS/DHCP server and I need to access devices like router or NAS with its reserved IP address.

What I would recommend is to keep only devices like routers, switches, DNS/DHCP servers, etc. which are part of network infra with a fixed IP address while let all other devices get dynamic addresses.  And for the rest of the devices, just don't bother with the IP address and use domain names that you can configure in the Scope. Using domain names for all devices on the network is much more convenient and easier to use.

> Thanks for listening!
> 
> Robin

You're welcome :)

**Reply by t1937r0** [reply]
So... I've configured multiple DHCP scopes as you had suggested.  Took a while. 

The first device I moved over to the new configuration was an M1 mac mini running Monterey 12.0.1.  
I had expected that the dhcp server would provide a domain search list including the dhcp scope's domain name (configured in the web ui for DnsServer.). It does not seem to do so.  Am I missing something?  Is there some other way to configure this that does not require changing the DNS configuration on the client machine?

If there is a more appropriate place to pose questions like this, please let me know.

Robin

**Reply by t1937r0** [reply]
Please ignore the previous post.  After I completely removed the wifi device from my mac's network configuration and then readded it, the domain search path populated correctly.

**Reply by ShreyasZare** [reply]
> Please ignore the previous post. After I completely removed the wifi device from my mac's network configuration and then readded it, the domain search path populated correctly.

Yes, that is normal since with DHCP its the client that requests the DHCP server to "pull" config. Server cant push changes to all clients. Clients usually will get new changes when they renew the lease, connect back to the network, or reboot.

**Comment by westdam** [comment]
What about a DHCP Relay or DHCP failover to another technitium? would be a nice feature.

**Reply by ShreyasZare** [reply]
Thanks for asking. DHCP failover will be covered with the DNS server clustering feature. DHCP relay agent is not planned since usually relay agent is available in routers where its needed.

**Comment by marco-belotti** [comment]
but wheeeeeen it will add DHCPv6?????? T_T

**Reply by ShreyasZare** [reply]
Its planned but due to time issues its taking a while.
